### PR TITLE
Chore/allow disabling public schema

### PR DESCRIPTION
**Pull request by saltcod**
depends on https://github.com/supabase/infrastructure/pull/17728

— 
public and storage schemas are currently protected (they can't be removed) 

This PR makes them removable, allowing project owners to essentially disable access to the public (and storage) schema.

screenshot-2024-04-10-at-09 11 56

Removing public schema gives you this alert:
screenshot-2024-04-10-at-14 49 33

With this alert on the api docs page, since it will no longer function properly:
screenshot-2024-04-10-at-14 50 13

### CODE REVIEW

File: apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx [hunk 1]
```diff
@@ -112,51 +126,60 @@ const PostgrestConfig = () => {
...
+
+                            {!isPublicSchemaEnabled && (
+                              <Alert_Shadcn_ variant="default">
+                                <Info className="h-4 w-4" />
+                                <AlertTitle_Shadcn_>
+                                  <ReactMarkdown>The `public` schema is not exposed</ReactMarkdown>
+                                </AlertTitle_Shadcn_>
+                                <AlertDescription_Shadcn_ className="flex flex-col gap-3">
+                                  <ReactMarkdown>
+                                    API endpoints will not be created for the `public` schema. You
```
Comment (inian) [inline apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx:170]: You will not be able to query tables and views in the public schema via postgREST and supabase-js.

File: apps/studio/components/interfaces/Settings/API/PublicSchemaNotEnabledAlert.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  const { ref: projectRef } = useParams()
+
+  return (
+    <>
+      <Alert_Shadcn_ variant="default">
+        <Info className="h-4 w-4" />
+        <AlertTitle_Shadcn_ className={context === 'docs' ? '!-mt-4' : ''}>
+          <ReactMarkdown>The `public` schema for this project is not exposed</ReactMarkdown>
+        </AlertTitle_Shadcn_>
+        <AlertDescription_Shadcn_ className={context === 'docs' ? 'grid gap-3 !-mt-7' : ''}>
```
Comment (saltcod) [inline apps/studio/components/interfaces/Settings/API/PublicSchemaNotEnabledAlert.tsx:21]: sorry, this nonsense required b/c docs page piles on its own custom css overriding (ruining) everything.

File: apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx [hunk 3]
```diff
@@ -65,15 +85,28 @@ const PostgrestConfig = () => {
...
+  const isPublicSchemaEnabled = config?.db_schema
+    .split(',')
+    .map((name) => name.trim())
+    .includes('public')
+
+  // const isGraphqlPublicSchemaEnabled = config?.db_schema
+  //   .split(',')
+  //   .map((name) => name.trim())
+  //   .includes('graphql_public')
+
```
Comment (alaister) [inline apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx:101]:
```suggestion
```

File: apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx [hunk 4]
```diff
@@ -65,15 +85,28 @@ const PostgrestConfig = () => {
...
+  const isGraphqlExtensionEnabled =
+    (extensions ?? []).find((ext) => ext.name === 'pg_graphql')?.installed_version !== null
+
   return (
     <Form id={formId} initialValues={initialValues} validate={() => {}} onSubmit={updateConfig}>
       {({ handleReset, resetForm, values, initialValues }: any) => {
         const hasChanges = JSON.stringify(values) !== JSON.stringify(initialValues)
-
+        console.log('schemas', config?.db_schema)
```
Comment (alaister) [inline apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx:109]:
```suggestion
```

File: apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx [hunk 5]
```diff
@@ -84,9 +117,11 @@ const PostgrestConfig = () => {
               max_rows: config.max_rows,
               db_extra_search_path: config.db_extra_search_path ?? '',
             }
+
             resetForm({ values, initialValues: values })
           }
         }, [config])
+        //}, [config])
```
Comment (alaister) [inline apps/studio/components/interfaces/Settings/API/PostgrestConfig.tsx:124]:
```suggestion
```

**Review by alaister — CHANGES_REQUESTED** [review]
Looking good!

There seems to be an issue where changing any of the items in the exposed schemas field submits the form:

https://github.com/supabase/supabase/assets/10985857/7d8d36fe-bcd8-4989-9ed1-8f4a83908433

Also a couple of small cleanup things:

**Review by alaister — APPROVED** [review]
We can fix the auto submission issue in a follow up PR 👍

**Comment by GaryAustin1** [comment]
Oh, goodie,  one more thing to check when things are going wrong.  Please warn them what a bad idea this is unless they know what they are doing.

**Pull request merged**